### PR TITLE
feat: support importing MD5 hashes

### PR DESCRIPTION
**Pull request by erolkskn**
## Related issue(s)
#2716 

## Checklist

- [x] I have read the [contributing guidelines](../blob/master/CONTRIBUTING.md).
- [x] I have referenced an issue containing the design document if my change introduces a new feature.
- [x] I am following the [contributing code guidelines](../blob/master/CONTRIBUTING.md#contributing-code).
- [x] I have read the [security policy](../security/policy).
- [ ] I confirm that this pull request does not address a security vulnerability. 
      If this pull request addresses a security. vulnerability, 
      I confirm that I got green light (please contact [security@ory.sh](mailto:security@ory.sh)) from the maintainers to push the changes.
- [x] I have added tests that prove my fix is effective or that my feature works.
- [ ] I have added or changed [the documentation](https://github.com/ory/docs).

### CODE REVIEW

File: hash/hash_comparator.go
Comment (aeneasr) [inline hash/hash_comparator.go:297]: Most md5 hashers probably use salting, but there's probably not one way to add the salt (as MD5 itself has no salting). How could we solve this?

Comment (erolkskn) [reply hash/hash_comparator.go:297]: Yeah that would be great addition. In most of the scenarios i have seen, salt is either appended or prepended to the password. And again probably not everyone uses salting. I think format like below, can accommodate up to that needs:
`$md5$<hash>$<salt base64>,<mode>`
- **mode** parameter can be either **a** *(abbreviation for append)* or **p** *(abbreviation for prepend)*. In the case of **a** salt will be appended to the password before it's being hashed. In the case of **p** salt will be prepended to the password before it's being hashed.
- The section after the third dollar sign (inclusive) is optional. That means if someone wants to store a password which has no salt, it can be formatted like this: `$md5$<hash>`. By putting **hash** before **salt** we can easily make *salting* optional.

Comment (erolkskn) [reply hash/hash_comparator.go:297]: In that case,
Password without a salt can look like this: `$md5$v=1$m=n$$202cb962ac59075b964b07152d234b70`
Password with salt can look like this: `$md5$v=1$m=a$MTIzNDU2Nw==$8a3379ae294935afa4195fa27600fd43`
Just wanted to make sure we are on the same page, am i getting it right ?

Comment (aeneasr) [reply hash/hash_comparator.go:297]: 🤔 I was also thinking that we could use sprintf to set the salt:

```
$md5$v=1$pf=%[0]s%[1]s$salt$hash
$md5$v=1$hash
```

where `pf` is `printf` and the first argument being the salt, the second the hash. If `pf` is not set, we do not have a salt. This would allow us to define the salt format any way we like:

```
%[0]s%[1]s
%[1]s%[0]s
%[0]sSOME_DELIMITER%[1]s
....
```

WDYT?

Comment (erolkskn) [reply hash/hash_comparator.go:297]: That's a good idea, this would make salting process more flexible. But i think formatting by index might come as *too implicit* for some users. How about formatting with named parameters like this: 
```
%[salt]s%[password]s
%[salt]sSOME_DELIMITER%[password]s
```
We might replace `%[salt]s` to `%[0]s` and `%[password]s` to `%[1]s` before passing format string to `sprintf`. For example format string `%[salt]s%[password]s` can turn into `%[0]s%[1]s` before passing it to `sprintf`. 
This way we both can use indexed formatting with sprintf and make the definition of format more explicit to the users.

What do you think ?

Comment (aeneasr) [reply hash/hash_comparator.go:297]: That's true, but I think it makes sense to just rely of `Sprintf`. In v2 we can think about introducing parsing here :)

Comment (erolkskn) [reply hash/hash_comparator.go:297]: Or we can make `$pf` variable enclosed by double quotes, and only thing the users would have to escape could be the double quotes itself and `%` char. 
For example:
```
<salt>[$<password> => $md5$v=1$pf="%[0]s[$%[1]s"$salt$hash
<salt>"%<password> => $md5$v=1$pf="%[0]s\"%%%[1]s"$salt$hash
```

Comment (erolkskn) [reply hash/hash_comparator.go:297]: As a third option we may not need any escaping at all. We can assume that part starts after `$pf=` and ends before `$salt` is the format string. For example:
```
<salt>[$<password> => $md5$v=1$pf=%[0]s[$$salt$hash
````
Even though that's an option, users would still need to escape `%` character, because of how Sprintf works.

Comment (erolkskn) [reply hash/hash_comparator.go:297]: > As a third option we may not need any escaping at all. We can assume that part starts after `$pf=` and ends before `$salt` is the format string. For example:
> 
> ```
> <salt>[$<password> => $md5$v=1$pf=%[0]s[$%[1]s$salt$hash
> ```
> 
> Even though that's an option, users would still have to escape `%` character, because of how Sprintf works.

We can capture the input with this regex: `\$md5\$v\=1\$pf\=(.*)\$(.*)\$(.*)`

Comment (aeneasr) [reply hash/hash_comparator.go:297]: I'm not sure if we have to escape it! Splitting by `$` and then processing each parameter in it's own context, it should work IMO :) The one problem I see is if the template itself uses `$`: `pf=%[0]s$%[1]s` 🤔 

In which case escaping it in double quotes might make sense: `pf="%[0]s$%[1]s"`

Comment (erolkskn) [reply hash/hash_comparator.go:297]: I think we are on different page now, so i'm just going to update the code according to thoughts i have in my mind, may be this can illuminate our conversation about the topic.

File: hash/hash_comparator.go
Comment (hperl) [inline hash/hash_comparator.go:312]: I think we are using base64 encoding everywhere else in password format strings. Why use hex encoding here?

**Review by hperl — COMMENTED** [review]
Hi, and thanks a lot for your PR. It looks really good already, I have just left some comments.

You can improve the coverage by adding more test cases to parse for formats that don't match.

File: hash/hasher_test.go
Comment (hperl) [inline hash/hasher_test.go:249]:
```suggestion
	assert.Error(t, hash.Compare(context.Background(), []byte("test"), []byte("$md5$v=1$5a105e8b9d40e1329780d62ea2265d8a")))
	assert.Error(t, hash.Compare(context.Background(), []byte("test"), []byte("$md5$v=1$ZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZ")))

```

Comment (hperl) [reply hash/hasher_test.go:249]: To improve test coverage ;)

File: hash/hash_comparator.go
Comment (erolkskn) [reply hash/hash_comparator.go:169]: It's based on the comment of Aeneas, see: https://github.com/ory/kratos/pull/2725#discussion_r968211034 It's just parameter for specifying which format version to use. It's not used for distinguishing salting from plain hash.

File: hash/hash_comparator.go
Comment (erolkskn) [reply hash/hash_comparator.go:312]: Honestly, I'm in between on this subject. In the one hand almost all of the products I've seen to this day keeps MD5 hashes in hexadecimal string format. On the other hand using hex encoding here contradicts with out formats as you say. My main point of concern in here is using base64 encoding might be a little confusing for some users. If we ask them to give it in base64 encoding instead of hex string, following questions might be arose:

- Should the user give base64 output of hex string itself ? 
- Should the user give base64 output of bytes itself ? 

Currently on other formats, we are using the strategy of encoding bytes with base64, so i'm assuming in that scenario we will go with the second option. In that case almost all of the users should convert hexadecimal formatted string to bytes first, and then they should encode those bytes with base64

File: hash/hash_comparator.go
Comment (erolkskn) [reply hash/hash_comparator.go:149]: When i took a look around, i hadn't seen any security issues regarding this. But i think this subject should be investigated in more depth.

File: hash/hasher_test.go
Comment (erolkskn) [reply hash/hasher_test.go:249]: Oh thanks for that :) I'm going to merge them.

File: hash/hash_comparator.go
Comment (hperl) [reply hash/hash_comparator.go]: I read the implementation yesterday and I still find the usage of the untrusted format string a little scary. At best it is way more powerful than we need it to be.

How about we just specify the format string as some string that contains `PASS` and `SALT` template strings, and then we `strings.Replace` the actual hash into that? The examples cover all use cases that I can think of (password or salt first, delimiters, etc.), e.g.:

* `PASS:SALT`
* `SALTPASS`
* `PASS$SALT`

To support literal `$`s in the template we can either base64-encode the whole string, or wrap the template with double quotes. I tend towards base64, because this is consistent with other fields in the format (salt and hash), and because the parsing format is then just a `strings.Split(s, "$")`, which is way easier to read than a regular expression.

WDYT?

File: hash/hash_comparator.go
Comment (hperl) [reply hash/hash_comparator.go:312]: Yes, I'd prefer that. The output of the MD5 hash itself should be base64 encoded.

File: hash/hash_comparator.go
Comment (hperl) [reply hash/hash_comparator.go:169]: So it is to distinguish the format string from some other thing we might want to introduce in the future?

If so, can't we worry about that then and introduce a new field?

So today we write
* `$md5$pf=some_v1_format_string$salt$hash`

If we want to change the format of `pf` in the future (if ever), we can introduce a new field, e.g.:

* `$md5$pfV2=some_v2_format_string$salt$hash`

The parser would ignore `pf` if there is also a `pfV2`.

Am I missing something? As I read it I found the version number somewhat confusing and didn't know what it related to, as MD5 itself does not have multiple versions.

File: hash/hash_comparator.go
Comment (hperl) [reply hash/hash_comparator.go:149]: I suggested a simpler format here: https://github.com/ory/kratos/pull/2725#discussion_r977279806

File: hash/hash_comparator.go
Comment (erolkskn) [reply hash/hash_comparator.go]: @hperl Definitely +1 on this. This feels way more secure than the `Printf` approach. About the base64 encoded format +1 on this too. This can ease the parsing process. Instead of using regex we can use `Split` approach as you suggested.

File: hash/hash_comparator.go
Comment (erolkskn) [reply hash/hash_comparator.go:169]: I think there is a little confusion here. The format is like the below in the current version:
`$md5$v=1$pf=...`
And if we introduce another version in the future, it's going to look like this:
`$md5$v=2$pf=...`

File: hash/hash_comparator.go
Comment (erolkskn) [reply hash/hash_comparator.go:312]: @hperl Alright then, just to be clear on this, users should be encoding 16 byte md5 hash with base64, right ? Not the 32 byte md5 hex string representation.

Comment (hperl) [reply hash/hash_comparator.go:312]: yes :)

File: hash/hash_comparator.go
Comment (hperl) [reply hash/hash_comparator.go:169]: My comment was that we don't need the version, and if in the future something changes we can version just that one field.

File: hash/hash_comparator.go
Comment (hperl) [reply hash/hash_comparator.go]: nice!

File: hash/hash_comparator.go
Comment (erolkskn) [reply hash/hash_comparator.go:169]: Oh, i think i got confused. I totally misread that, sorry about this. Yeah it's probably more convenient to do :-)

**Review by hperl — DISMISSED** [review]
Thanks @erolkskn, LGTM! :tada: 
@zepatrik can we merge this now or should we wait for final approval by @aeneasr?

/cc @jonas-jonas

**Review by zepatrik — APPROVED** [review]
Looks good IMO. We will merge it beginning of next week :+1:

**Review by aeneasr — DISMISSED** [review]

**Comment by aeneasr** [comment]
Awesome! This is mergable!

The last missing piece is no adding a bit of docs for this feature :) https://www.ory.sh/docs/kratos/manage-identities/import-user-accounts-identities#bcrypt-pkbdf2-argon2-scrypt-family-hashed-password

We probably need a dedicated section for MD5 as our implementation is not standardized, meaning that we'll need to explain how this works :)

**Review by hugotiburtino — APPROVED** [review]
Beautiful!

**Comment by erolkskn** [comment]
Oh sorry for this took too long. I couldn't find appropriate time to write down docs. It should be ready in the next few days. Going to rebase with master tonight.

**Comment by aeneasr** [comment]
Awesome! And there have since been a few merge conflicts - could you please also take a look into those? :)

**Comment by erolkskn** [comment]
> Awesome! And there have since been a few merge conflicts - could you please also take a look into those? :)

Of course :-) Going to look at them as soon as I can today

**Comment by aeneasr** [comment]
Thank you! We still need to explain this feature in the documentation. Do you need help for that? :)

**Comment by erolkskn** [comment]
> Thank you! We still need to explain this feature in the documentation. Do you need help for that? :)

I'm on it now :-) I'm going to create a PR regarding this on "ory/docs" this evening.

**Comment by aeneasr** [comment]
awesome!!!

**Review by hperl — APPROVED** [review]
LGTM! Thanks for the good work, @erolkskn!

**Comment by graysontvbs** [comment]
hi , I am using the "import" command to import the identity, and the password is encrypted using the md5 method. However, I cannot log in correctly. After checking, I found that if the previous password was encrypted using a different programming language, it cannot be properly compared and verified.

This is because the md5 encryption method in Go differs from other programming languages (Go uses bytes). However, in older systems, different programming languages such as PHP or C# may have been used. This prevents me from using those md5 passwords for login.

In summary, it seems that there are two different md5 encryption methods, and it appears that they are not supported at the moment. Can we open a separate issue to address this matter?

**Comment by jonas-jonas** [comment]
>I found that if the previous password was encrypted using a different programming language, it cannot be properly compared and verified.

With which programming language were your passwords hashed before? In theory the implementation in Kratos should follow the spec, but if the passwords were hashed with some custom implementation, that deviates from the spec, Kratos won't be able to verify the passwords.

Please also make sure, [to include the salt](https://www.ory.sh/docs/kratos/manage-identities/import-user-accounts-identities#hashed-passwords) if a salt was used to hash the passwords.

>Can we open a separate issue to address this matter? 

Please do!

**Comment by graysontvbs** [comment]
I think I have reached some conclusions. The unit test cases for this PR are incorrect. Taking the first test case as an example, after encrypting with md5 and base64, the expected result should be `MDk4ZjZiY2Q0NjIxZDM3M2NhZGU0ZTgzMjYyN2I0ZjY=`, but the test case uses `CY9rzUYh03PK3k6DJie09g==`.

The reason for this is that the encryption process involves passing the byte[:] result of md5 into the base64 encode function, rather than passing the string result of md5. If you use the following sample code, you can clearly see that the base64 output will be different depending on the input data:


```go
md5Hash := md5.Sum([]byte("test")) // 098f6bcd4621d373cade4e832627b4f6

//use byte[:] to base64 encode (X)
base64Encoded := base64.StdEncoding.EncodeToString(md5Hash[:])
fmt.Println(base64Encoded) //CY9rzUYh03PK3k6DJie09g==

//use string to base64 encode (O)
base64Encoded1 := base64.StdEncoding.EncodeToString([]byte(hex.EncodeToString(md5Hash[:])))
fmt.Println(base64Encoded1) //MDk4ZjZiY2Q0NjIxZDM3M2NhZGU0ZTgzMjYyN2I0ZjY=
```

**Pull request merged**
